### PR TITLE
bpo-45847: Port _tkinter to PY_STDLIB_MOD

### PR DESCRIPTION
**Pull request by erlend-aasland**
- Remove ``--with-tclk-*`` options from `configure`
- Use pkg-config to detect `_tkinter` dependencies (Tcl/Tk, X11)
- Manual override via environment variables `TCLTK_CFLAGS` and `TCLTK_LIBS`

https://bugs.python.org/issue45847

Automerge-Triggered-By: GH:tiran

### CODE REVIEW

**Comment by erlend-aasland** [comment]
Looks like the Ubuntu pkg-config files for Tcl/Tk are buggy; they don't provide the correct `--libs` flags (`-L` is wrong).

**Comment by erlend-aasland** [comment]
(I squeezed 46e43dcaa599d3fe906cdb3ebd591151af856372, 5156b5764a368578e5eb55da471b74e80e5aa479, and 8af1b258d04b042038e0cc0e5fccfd35aad659af into 72cf59205dd1a16b949a85ee0967ecc32d6a85cc)

File: configure.ac [hunk 1]
```diff
@@ -3499,38 +3499,73 @@ AS_VAR_IF([enable_loadable_sqlite_extensions], [yes], [
...
+
+  AC_LINK_IFELSE([
+    AC_LANG_PROGRAM([
+      #include <tcl.h>
+      #include <tk.h>
+      #if defined(TK_HEX_VERSION)
+      #  if TK_HEX_VERSION < 0x08040200
+      #    error "Tk older than 8.4.2 not supported"
+      #  endif
+      #endif
```
Comment (erlend-aasland) [inline configure.ac]: Drop this and keep the ones below?

**Comment by erlend-aasland** [comment]
Buildbot status embedded below (updating this post at irregular intervals)

<details>
<summary>Buildbot status</summary>

Buildbot | Status | Comment
--- | --- | ---
AMD64 Debian PGO PR | 🟢
AMD64 Debian root PR | 🟢
AMD64 RHEL8 FIPS Only Blake2 Builtin Hash PR  | 🟢 
AMD64 Ubuntu Shared PR | 🟢 
ARM64 macOS PR | 🔴 | missing (change of behaviour)
PPC64 Fedora PR | 🔴 | missing (change of behaviour)
s390x Debian PR | 🟢 
s390x SLES PR | 🟢 
x86 Gentoo Installed with X PR  | 🟢 
x86 Gentoo Non-Debug with X PR | 🟢 
x86-64 macOS PR | 🔴 | missing (change of behaviour)

</details>

**Comment by erlend-aasland** [comment]
Sparked by [Ronald's comment on Discourse](https://discuss.python.org/t/simplifying-tcl-tk-detection/14066/7?u=erlendaasland):

We _could_ add a hack for macOS... I added this on my macOS 12 install, and it works for me. Not sure how it would fare on older installs, though.

```diff
--- a/configure.ac
+++ b/configure.ac
@@ -3515,8 +3515,21 @@ for _QUERY in \
 done
 
 AS_VAR_IF([found_tcltk], [no], [
-  TCLTK_CFLAGS=${TCLTK_CFLAGS-""}
-  TCLTK_LIBS=${TCLTK_LIBS-""}
+  AS_CASE([$ac_sys_system],
+    [Darwin*], [
+      TCLTK_CFLAGS=${TCLTK_CFLAGS-"\
+        -I/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/System/Library/Frameworks/Tcl.framework/Headers \
+        -I/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/System/Library/Frameworks/Tk.framework/Headers \
+        -F/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/System/Library/Frameworks \
+      "}
+      TCLTK_LIBS=${TCLTK_LIBS-"\
+        -Wl,-F,/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/System/Library/Frameworks,-framework,Tcl,-framework,Tk
+      "}
+    ], [
+      TCLTK_CFLAGS=${TCLTK_CFLAGS-""}
+      TCLTK_LIBS=${TCLTK_LIBS-""}
+    ]
+  )
 ])
 
 dnl FreeBSD has an X11 dependency which is not implicitly resolved.
```

I'm not sure if it's worth it adding a more complex hack, or if it's worth it adding this hack in the first place.

cc. @tiran @ronaldoussoren @ned-deily

**Comment by erlend-aasland** [comment]
> We _could_ add a hack for macOS... I added this on my macOS 12 install, and it works for me. Not sure how it would fare on older installs, though.

Based on comments from Ned and Ronald, I'm dropping adding any hack for macOS; those building on macOS will have to install MacPorts or Homebrew, and explicitly or implicitly pass this information to `configure`, for example through `PKG_CONFIG_PATH`.

**Comment by erlend-aasland** [comment]
Post buildbot run: There are some pdb and distutils failures. I recon they are unrelated.

**Comment by erlend-aasland** [comment]
Updated to reduce configure noise.

**Comment by erlend-aasland** [comment]
We can now remove `macosx_sdk_specified` and `find_library_file` from  `setup.py`. I'll push that; sorry for touching the PR again :)

**Comment by erlend-aasland** [comment]
Synced with main to get Serhiy's commits (GH-31839, GH-31938)

**Comment by erlend-aasland** [comment]
@tiran, do you want another round through the buildbots after I pulled Serhiy's changes? Apart from that, are you fine with this PR as it stands?

**Comment by ned-deily** [comment]
Thanks for the PR!  I'd like to review and play with it a bit but I likely won't be able to do so before Monday. If you don't hear anything by then, you go ahead.

**Comment by erlend-aasland** [comment]
@ned-deily would you still like to review this?

File: Doc/whatsnew/3.11.rst [hunk 2]
```diff
@@ -693,8 +693,8 @@ Build Changes
 
 * Build dependencies, compiler flags, and linker flags for most stdlib
   extension modules are now detected by :program:`configure`. libffi, libnsl,
-  libsqlite3, zlib, bzip2, liblzma, libcrypt, and uuid flags are detected by
-  ``pkg-config`` (when available).
+  libsqlite3, zlib, bzip2, liblzma, libcrypt, Tcl/Tk libs, and uuid flags
+  are detected by ``pkg-config`` (when available).
```
Comment (ned-deily) [inline Doc/whatsnew/3.11.rst:774]: Somewhere  in here, we should explicitly document the removal of the `--with-tcltk-includes` and `--with-tcltk-libs` configuration options. I have seen them (and the corresponding Makefile variables) in use in the wild.

**Review by ned-deily — COMMENTED** [review]
LGTM, thanks for doing this. Considering the history and complexity, I'm sure some problems will pop up that, at a minimum, will require more documentation, including the devguide with regard to installing dependencies. I did leave a comment about explicitly documenting the removal of the previous `configure` options.  Also FTR, the macOS installer build script at the moment still depends on those current configure options; I will rework that area prior to the next 3.11 alpha.

File: Doc/whatsnew/3.11.rst (hunk 2 above)
Comment (erlend-aasland) [reply Doc/whatsnew/3.11.rst:774]: I agree. I'll add a sentence or two to the build section of What's New. I'll also create a PR for the devguide.

Comment (erlend-aasland) [reply Doc/whatsnew/3.11.rst:774]: Is b48d671fe1330933322f715e3aabdb234533f38a sufficient?

**Comment by ned-deily** [comment]
Ready to go

**Comment by erlend-aasland** [comment]
> Ready to go

Great, thank you for reviewing! @tiran, can you land this? :)

**Review by tiran — APPROVED** [review]

**Comment by tiran** [comment]
Love it! Your PR removes a lot of complicated code from ``setup.py``.

**Comment by sharewax** [comment]
files from centos7 (tk-devel-8.5.13-6.el7.x86_64 and tcl-devel-8.5.13-8.el7.x86_64) doesn't contain files for pkg-config, so tcl for debug version of python you get
```gcc -pthread -shared -L/opt/wargaming/wargaming-python311/root/usr/lib64-Wl,-z,relro -Wl,-rpath,/opt/wargaming/wargaming-python311/root/usr/lib64 -
Wl,--enable-new-dtags -g -L/usr/lib64/openssl11 -L/opt/wargaming/wargaming-python311/root/usr/lib64-Wl,-z,relro -Wl,-rpath,/opt/wargaming/wargaming
-python311/root/usr/lib64 -Wl,--enable-new-dtags -g -L/usr/lib64/openssl11 -L/opt/wargaming/wargaming-python311/root/usr/lib64-Wl,-z,relro -Wl,-rpa
th,/opt/wargaming/wargaming-python311/root/usr/lib64 -Wl,--enable-new-dtags -fno-semantic-interposition -g -L/usr/lib64/openssl11 -flto -fuse-linke
r-plugin -ffat-lto-objects -flto-partition=none -g -L/opt/wargaming/wargaming-python311/root/usr/lib64-Wl,-z,relro -Wl,-rpath,/opt/wargaming/wargam
ing-python311/root/usr/lib64 -Wl,--enable-new-dtags -fno-semantic-interposition -g -L/usr/lib64/openssl11 -L/opt/wargaming/wargaming-python311/root
/usr/lib64-Wl,-z,relro -Wl,-rpath,/opt/wargaming/wargaming-python311/root/usr/lib64 -Wl,--enable-new-dtags -g -L/usr/lib64/openssl11 -I/opt/wargami
ng/wargaming-python311/root/usr/include -O2 -g -pipe -Wall -Wp,-D_FORTIFY_SOURCE=2 -fexceptions -fstack-protector-strong --param=ssp-buffer-size=4
-grecord-gcc-switches -m64 -mtune=generic -D_GNU_SOURCE -fPIC -fwrapv -I/usr/include/openssl11 build/temp.linux-x86_64-3.11-pydebug/builddir/build/
BUILD/Python-3.11.0rc1/Modules/_ctypes/_ctypes.o build/temp.linux-x86_64-3.11-pydebug/builddir/build/BUILD/Python-3.11.0rc1/Modules/_ctypes/callbac
ks.o build/temp.linux-x86_64-3.11-pydebug/builddir/build/BUILD/Python-3.11.0rc1/Modules/_ctypes/callproc.o build/temp.linux-x86_64-3.11-pydebug/bui
lddir/build/BUILD/Python-3.11.0rc1/Modules/_ctypes/cfield.o build/temp.linux-x86_64-3.11-pydebug/builddir/build/BUILD/Python-3.11.0rc1/Modules/_cty
pes/stgdict.o -L. -L/opt/wargaming/wargaming-python311/root/usr/lib64 -L/usr/lib64/openssl11 -L/usr/local/lib -lffi -ldl -o build/lib.linux-x86_64-
3.11-pydebug/_ctypes.cpython-311d-x86_64-linux-gnu.so
*** WARNING: renaming "_tkinter" since importing it failed: /builddir/build/BUILD/Python-3.11.0rc1/build/debug/build/lib.linux-x86_64-3.11-pydebug/_tkinter.cpython-311d-x86_64-linux-gnu.so: undefined symbol: Tcl_AddErrorInfo
Following modules built successfully but were removed because they could not be imported:
_tkinter
```

I'll try to set manually flags TCLTK_CFLAGS and TCLTK_LIBS, but it seems strage. on python3.10 all works well without additional things.

**Comment by tiran** [comment]
@sharewax Python 3.11 requires pkg-config to detect and configure tkinter. The CentOS 7 packages do not provide a ``.pc`` file for TCL and TK. `_tkinter` compiles for me with correct `TCLTK_LIBS`. We should document the correct flags for CentOS7 users. Could you please open a new ticket?

**Comment by sharewax** [comment]
@tiran Sure. Can you provide correct `TCLTK_LIBS` for centos7? I would like to finish my installation correctly :)

**Pull request merged**